tqdm/std.py

```diff
@@ -1271,41 +1283,45 @@ def close(self):
         # Prevent multiple closures
         self.disable = True
 
-        # decrement instance pos and remove from internal set
+        # decrement instance pos
         pos = abs(self.pos)
-        self._decr_instances(self)
 
-        if self.last_print_t < self.start_t + self.delay:
-            # haven't ever displayed; nothing to clear
-            return
+        try:
+            if self.last_print_t < self.start_t + self.delay:
+                # haven't ever displayed; nothing to clear
+                return
 
-        # GUI mode
-        if getattr(self, 'sp', None) is None:
-            return
+            # GUI mode
+            if getattr(self, 'sp', None) is None:
+                return
 
-        # annoyingly, _supports_unicode isn't good enough
-        def fp_write(s):
-            self.fp.write(str(s))
+            # annoyingly, _supports_unicode isn't good enough
+            def fp_write(s):
+                self.fp.write(str(s))
 
-        try:
-            fp_write('')
-        except ValueError as e:
-            if 'closed' in str(e):
-                return
-            raise  # pragma: no cover
+            try:
+                fp_write('')
+            except ValueError as e:
+                if 'closed' in str(e):
+                    return
+                raise  # pragma: no cover
 
-        leave = pos == 0 if self.leave is None else self.leave
+            leave = pos == 0 if self.leave is None else self.leave
 
-        with self._lock:
-            if leave:
-                # stats for overall rate (no weighted average)
-                self._ema_dt = lambda: None
-                self.display(pos=0)
-                fp_write('\n')
-            else:
-                # clear previous display
-                if self.display(msg='', pos=pos) and not pos:
-                    fp_write('\r')
+            with self._lock:
+                if leave:
+                    # stats for overall rate (no weighted average)
+                    self._ema_dt = lambda: None
+                    self.display(pos=0)
+                    fp_write('\n')
+                else:
+                    # clear previous display
+                    if self.display(msg='', pos=pos) and not pos:
+                        fp_write('\r')
+
+        finally:
+            # Remove from internal set
+            self._decr_instances(self)
```

Comment:
I moved this to a finally block to *first* adjust the closed bar to the top row, before other bars adjust their position upwards. This results in a cleaner screen update, I think.